contracts/Thecosomata.sol

```diff
@@ -0,0 +1,332 @@
+
+        return debtLimit.sub(debtBalance);
+    }
+
+    /**
+        @notice Get the amount of sOHM that can be withdrawn and unstaked
+        @return uint256 Unstakeable sOHM
+     */
+    function getRemainingUnstakeableSOHM() internal view returns (uint256) {
+        uint256 totalSOHM = IsOHM(sOHM).balanceOf(RedactedTreasury).add(
+            IsOHM(sOHM).balanceOf(address(this))
+        );
+        uint256 sOHMReservedForOlympus = IOlympusTreasury(OlympusTreasury)
+            .debtLimit(address(this))
+            .mul(2);
+
+        return totalSOHM.sub(sOHMReservedForOlympus);
+    }
+
+    /**
+        @notice Withdraw sOHM from Redacted treasury
+        @param  amount uint256 The amount of sOHM to withdraw
+     */
+    function withdrawSOHMFromTreasury(uint256 amount) internal {
+        IRedactedTreasury(RedactedTreasury).manage(sOHM, amount);
+    }
+
+    /**
+        @notice Borrow OHM from Olympus treasury using uncollateralized sOHM
+     */
+    function incurOlympusDebt(uint256 amount) internal {
+        IOlympusTreasury(OlympusTreasury).incurDebt(amount, OHM);
+    }
+
+    /**
+        @notice Add OHM and BTRFLY as liquidity to Sushi LP
+        @param ohmAmount    uint256 OHM to add as liquidity
+        @param btrflyAmount uint256 BTRFLY to add as liquidity
+     */
+    function addOHMBTRFLYLiquiditySushiSwap(
+        uint256 ohmAmount,
+        uint256 btrflyAmount
+    ) internal {
+        (, , uint256 slpMinted) = ISushiRouter(SushiRouter).addLiquidity(
+            BTRFLY,
+            OHM,
+            btrflyAmount,
+            ohmAmount,
+            btrflyAmount,
+            ohmAmount,
+            address(this), // Mint LP tokens directly to Redacted treasury
+            block.timestamp + 5 minutes
+        );
+
+        IERC20 slpContract = IERC20(
+            ISushiFactory(sushiFactory).getPair(OHM, BTRFLY)
+        );
+
+        uint256 olympusFee;
+        uint256 redactedDeposit;
+
+        if (debtFee > 0) {
+            // Send Olympus fee in the form of LP tokens
+            olympusFee = slpMinted.mul(debtFee).div(1000000);
+
+            slpContract.transfer(OlympusTreasury, olympusFee);
+        }
+
+        redactedDeposit = slpMinted.sub(olympusFee);
+
+        // Transfer LP token balance to Redacted treasury
+        slpContract.transfer(RedactedTreasury, redactedDeposit);
+
+        emit TransferLPTokens(olympusFee, redactedDeposit);
+    }
+
+    /**
+        @notice Unstake sOHM balance
+     */
+    function unstakeSOHM(uint256 amount) internal {
+        IOlympusStaking(OlympusStaking).unstake(
+            address(this),
+            amount,
+            true,
+            true
+        );
+    }
+
+    /**
+        @notice Borrow from the Olympus Treasury and add liquidity
+        @param  shouldBorrow bool Whether we should borrow or unstake
+     */
+    function addLiquidity(bool shouldBorrow) internal {
+        uint256 btrfly = IBTRFLY(BTRFLY).balanceOf(address(this));
```

Comment:
Need to revert if zero

Reply:
Already worked on the refactoring based on the changes in the CRV one, including more checkpoints. Will push up a new PR instead and close this one later once we merged the PR for CRV one 👍